### ISSUE TITLE
How can i train my own model?

### ISSUE THREAD

**Issue by fjzpcmj** [cguindel/lsi-faster-rcnn#2]
How can i train my own model?  I'll appreciate it !

**Comment by cguindel** [comment]
The original py-faster-rcnn allows training using Pascal VOC 2007/2010/2012 and COCO, and I have modified it to include the KITTI object dataset, where viewpoint labels are available. To train using KITTI:

1. Create a training script in experiments/scripts, similar to leaderboard.sh.
2. Add a configuration file with the training parameters in experiments/cfgs; check leaderboard.yml for example.
3. Create a model folder in models/kitti/{VGG16,ZF}/ with a solver and its train/test prototxt files.

Then, run the training script. If you run leaderboard.sh as it is, it will train my model to get the official results [at here](http://www.cvlibs.net/datasets/kitti/eval_object_detail.php?&result=265539b97ba7875309d064e6f3e380a0eb7b3a89). Trained networks are saved under output/{experiment directory}/{dataset name}/. If you need to evaluate the results, you can create a test_*.sh script in experiments/scripts; then you can use [this](https://github.com/cguindel/eval_kitti) to obtain the KITTI official metrics. 

In case you want to use your own dataset, the process is unfortunately not so straightforward, and you will probably need to understand the code and modify it, particularly the one placed in lib/datasets. You can get some help in the official py-faster-rcnn repository; e.g. [this issue](https://github.com/rbgirshick/py-faster-rcnn/issues/243). Another option is just adapting the annotations of the new dataset to follow the format of the KITTI (or Pascal, or COCO) annotations and "fake" its folder structure. Please keep in mind that my modifications should be dataset-agnostic, but I have not actually tested them outside the KITTI dataset.

If you provide more details about what you are trying to do, maybe I can further help you.

**Comment by fjzpcmj** [comment]
Hi @cguindel , thanks for your detailed answer sincerely.I use train_kitti_lsi_frcnn.sh to retrain the model.Is there any difference between leaderboard.sh and train_kitti_lsi_frcnn.sh? Can i use test_kitti_lsi_frcnn.sh to test my model? I am a starter in object detection,first i just want to finish my class project about car detection and viewpoint estimation.Again thanks very much.

**Comment by cguindel** [comment]
train_kitti_lsi_frcnn.sh is a former version which does not include some improvements that I've discovered later.  I would strongly recommend using leaderboard.sh, if you want to detect Car/Pedestrian/Cyclist or leaderboard_7cls.sh, to train the model using all the available classes (Person_sitting, Tram, etc.)

Do you want to obtain quantitative results (i.e., Average Precision, etc.)? If not, you do not need to use a test script. Otherwise, you will need to split the training set into train/validation subsets. To that end, you will need to modify this line in leaderboard.sh:
`TRAIN_IMDB="kitti_training"`
to:
`TRAIN_IMDB="kitti_trainsplit"`
and then use test_leaderboard.sh replacing:
```
TRAIN_IMDB="kitti_training"
TEST_IMDB="kitti_test_testing"
```
by
```
TRAIN_IMDB="kitti_trainsplit"
TEST_IMDB="kitti_valsplit"
```
That way, you will be training with 3682 images and obtaining results based on 3799 validation images.  I think this is the only change that you will need to undertake in order to re-train the model after the last commit 63dbdc3. Please try it and tell me what you got; once you have the model, I can help you to obtain detection stats if you need it.

**Comment by fjzpcmj** [comment]
It's very nice of you to help me so much.My project only need to detect car and car's viewpoint.I want to change your code to classify only two class(car and background).Also,have you ever used continuous viewpoint estimation,that is to say, consider the problem as  a regression problem..First,I should read your code carefully.Then I will try my best to do some change to fit my project.I will be grateful if you can help me to obtain quantitative results.

**Comment by cguindel** [comment]
To modify the number of classes, you need to change the parameter CLASSES in the configuration file under experiments/cfgs/ (in your case, you should set it to `CLASSES: ['__background__', 'Car']`) and the train/test prototxt files under models/. For the latter, I tried to add a comment just above every instance that depends on the number of classes; I recommend searching for `N_CLASSES` and modify the corresponding values from 4 to 2.

Regarding the continuous viewpoint estimation: yes, I am currently trying different alternatives (e.g., regression using Smooth L1 loss), but I have found that they do not improve the results, at least in terms of Average Orientation Similarity (the metric used in the KITTI benchmark). I plan to release the code to replicate these experiments, though; but that is a long-term project. 

If you finally achieve some results using this code, please let me know!

**Comment by fjzpcmj** [comment]
Should I change the **matrix H**?because a problem occurs that 
_in infogain_loss_layer.cpp line 72 : infogain->count()==num_labels_*num_lables_ (16 vs 4)_
,it seems that I should make the matrix H for my case.Following your paper, I find the formulation of matrix H, but how can i get f_k and f_min?

**Comment by cguindel** [comment]
Oh, yes, sorry. I forgot about the infogain matrix. As you may have read in the paper, I use the infogain loss to tackle the effect of imbalanced classes in the dataset; particularly, the presence of more cars than pedestrians and cyclists. If you are only using the 'Car' label, you probably can replace the infogain loss with a regular 'SoftmaxWithLoss' loss layer. If you still want to manage the imbalance between 'Car' and background samples, you will need to edit the infogainH.binaryproto; the relevant values that I use in the H matrix are ![h00](https://user-images.githubusercontent.com/15200248/34358046-7b38aa9a-ea4c-11e7-8b46-6744700ce84e.gif) for background regions and ![h11](https://user-images.githubusercontent.com/15200248/34358055-869605f4-ea4c-11e7-83af-bc94837d019e.gif) for cars. Anyway, note that equation 9 from the paper cannot be applied here because you only have a class.

**Comment by ahmed-shariff** [comment]
For something like this would it not be convenient if the `models` (faster_rcnn, rpn, etc.) have documentations on the format of the input data expected? and maybe a quick demo on how they can be used?
I was trying to include the faster-rcnn as a module in one of my own networks and play around with it. Haven't made much progress on it though, still trying to understand how some parts work.

**Comment by cguindel** [comment]
I am closing this issue due to the lack of activity.
